tensorflow_probability/python/distributions/empirical.py

```diff
@@ -0,0 +1,260 @@
+        is based. The first `k` dimensions index into a batch of independent
+        distributions. Length of `S` dimension determines number of samples
+        in each multiset. The last `n` dimension represents samples for each
+        distribution. n is specified by argument event_dims.
+      event_ndims: Python `int32`, default `0`. number of dimensions for each
+        event. When `0` this distribution has scalar samples. When `1` this
+        distribution has vector-like samples.
+      validate_args: Python `bool`, default `False`. When `True` distribution
+        parameters are checked for validity despite possibly degrading runtime
+        performance. When `False` invalid inputs may silently render incorrect
+        outputs.
+      allow_nan_stats: Python `bool`, default `True`. When `True`, statistics
+        (e.g., mean, mode, variance) use the value "`NaN`" to indicate the
+        result is undefined. When `False`, an exception is raised if one or
+        more of the statistic's batch members are undefined.
+      name: Python `str` name prefixed to Ops created by this class.
+
+    Raises:
+      ValueError: if the rank of `samples` < event_ndims+1.
+    """
+
+    parameters = locals()
+    with tf.name_scope(name, values=[samples]):
+      self._samples = tf.convert_to_tensor(samples, name="samples")
+      self._event_ndims = event_ndims
+      with tf.control_dependencies(
+          [tf.assert_rank_at_least(self._samples, event_ndims+1)]):
+        try:
+          samples_shape = util.prefer_static_shape(self._samples)
+          self._num_samples = samples_shape[-event_ndims-1]
+        except IndexError:
+          self._num_samples = tf.constant(1)
+
+    super(Empirical, self).__init__(
+        dtype=self._samples.dtype,
+        reparameterization_type=reparameterization.FULLY_REPARAMETERIZED,
+        validate_args=validate_args,
+        allow_nan_stats=allow_nan_stats,
+        parameters=parameters,
+        graph_parents=[self._samples],
+        name=name)
+
+  @staticmethod
+  def _param_shapes(sample_shape):
+    return {"samples": tf.convert_to_tensor(sample_shape, dtype=tf.int32)}
+
+  @property
+  def samples(self):
+    """Distribution parameter."""
+    return self._samples
+
+  @property
+  def num_samples(self):
+    """Number of samples."""
+    return self._num_samples
+
+  def _batch_shape_tensor(self):
+    return tf.shape(self.samples)[:-self._event_ndims-1]
+
+  def _batch_shape(self):
+    return self.samples.shape[:-self._event_ndims-1]
+
+  def _event_shape_tensor(self):
+    if self._event_ndims == 0:
+      return tf.constant([], dtype=tf.int32)
+    return tf.shape(self.samples)[-self._event_ndims:]
+
+  def _event_shape(self):
+    if self._event_ndims == 0:
+      return tf.TensorShape([])
+    return self.samples.shape[-self._event_ndims:]
+
+  def _mean(self):
+    return tf.reduce_mean(self.samples, axis=-self._event_ndims-1)
```

Comment:
Agreed that introducing `self._samples_axis` will make it easier. But `self.samples.shape[:self._samples_axis]` raises a TypeError for slicing samples with dynamic shape:

----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/johnzyh/PycharmProjects/probability/tensorflow_probability/python/distributions/empirical_test.py", line 63, in testShapes
    self.evaluate(dist.batch_shape_tensor()))
  File "/home/johnzyh/PycharmProjects/probability/tensorflow_probability/python/distributions/distribution.py", line 572, in batch_shape_tensor
    if self.batch_shape.is_fully_defined():
  File "/home/johnzyh/PycharmProjects/probability/tensorflow_probability/python/distributions/distribution.py", line 593, in batch_shape
    return tf.TensorShape(self._batch_shape())
  File "/home/johnzyh/PycharmProjects/probability/tensorflow_probability/python/distributions/empirical.py", line 176, in _batch_shape
    return self.samples.shape[:self._samples_axis]
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/framework/tensor_shape.py", line 806, in __getitem__
    elif start < 0 or stop < 0:
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/framework/ops.py", line 657, in __bool__
    raise TypeError("Using a `tf.Tensor` as a Python `bool` is not allowed. "
TypeError: Using a `tf.Tensor` as a Python `bool` is not allowed. Use `if t is not None:` instead of `if t:` to test if a tensor is defined, and use TensorFlow ops such as tf.cond to execute subgraphs conditioned on the value of a tensor.

I also tried `tf.slice()`, but it returns a tensor rather than a TensorShape.

Reply:
I actually failed to use `self.samples.shape[:self._samples_axis]` in `_batch_shape()` and `_event_shape()`. A TensorShape is required here but tf.shape returns a tensor. I don't see how to circumvent that.

Reply:
I fixed it.